cephfs-shell: Fix rmdir -p issues and add tests for rmdir

varshar16 opened the pull request:
This patch series fixes following issues:
- Deletion of non-empty directories
- Deletion of only last child directory when root directories are also passed

Also adds tests for rmdir.

Fixes: https://tracker.ceph.com/issues/40861
Fixes: https://tracker.ceph.com/issues/40863
Signed-off-by: Varsha Rao <varao@redhat.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->


<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+            o = self.mount_a.stat('test_dir')
+            log.info("mount_a output:\n{}".format(o))
+            raise Exception("Something went wrong!! test_dir is not deleted")
+        except CommandFailedError:
```
Looking for `CommandFailedError` is not a nice way. Often, other kind of failures leading to `CommandFailedError` gets misinterpreted as valid failure.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+            o = self.mount_a.stat('test_dir')
+            log.info("mount_a output:\n{}".format(o))
+            raise Exception("Something went wrong!! test_dir is not deleted")
+        except CommandFailedError:
```
I'll make same point about `CommandFailedError`

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+            o = self.mount_a.stat('test_dir')
+            log.info("mount_a output:\n{}".format(o))
+            raise Exception("Something went wrong!! test_dir/t1/t2/t3 are not deleted")
+        except CommandFailedError:
```
same point.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        o = self.run_cephfs_shell_cmd("rmdir test_dir")
+        print_output = o.stderr.getvalue()
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('Directory not empty') != -1
```
I would prefer checking for failure and `stat`ing for presence of `test_dir`. This way we would also test a bug where the method might delete the underlying file or the directory itself and yet somehow ends with valid failure.

Depending on error message is not the best, IMO, since minor change in it would make updating tests mandatory.

And if you do want match test, use case-insensitive search or setting everything to lowercase can prevent scope of irritation in future.

rishabh-d-dave replied:
> Depending on error message is not the best, IMO, since minor change in it would make updating tests mandatory.

Happened to me today, test_snap is broken -
```
CephFS:~/>>> snap create snap1 /data_dir
snapshot 'snap1' already exists
CephFS:~/>>>
```

From https://github.com/ceph/ceph/blob/master/qa/tasks/cephfs/test_cephfs_shell.py#L346 -
```
 assert(o[0] == "ERROR: snapshot 'snap1' already exists")
```

varshar16 replied:
You are right, I will change it.

rishabh-d-dave replied:
Cool.

Also, [uniittest provides a large list of assert methods](https://docs.python.org/3.7/library/unittest.html#assert-methods), use those instead of the `assert` keyword since it leads to a better output in case of a failure.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        o = self.get_cephfs_shell_cmd_output("mkdir -p test_dir/t1/t2/t3")
+        log.info("cephfs-shell output:\n{}".format(o))
+
+        o = self.run_cephfs_shell_cmd("rmdir test_dir")
```
Preferably use `p` or proc instead of  `o`.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        o = self.run_cephfs_shell_cmd("rmdir test_dir")
+        print_output = o.stderr.getvalue()
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('Directory not empty') != -1
```
same.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        except CommandFailedError:
+            pass
+
+    def test_rmdir_empty_sub_dir(self):
```
perhaps, change title to reflect that directory does contain sub-directory.

rishabh-d-dave replied:
s/`test_rmdir_empty_sub_dir`/`test_rmdir_on_dir_containing_subdir`

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('Directory not empty') != -1
+
+    def test_rmdir_non_empty_dir(self):
```
`test_rmdir_non_empty_dir`/`test_rmdir_on_dir_containing_file`

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        o = self.run_cephfs_shell_cmd("rmdir test_dir")
+        print_output = o.stderr.getvalue()
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('No such file or directory') != -1
```
same.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('No such file or directory') != -1
+
+    def test_rmdir_valid_file(self):
```
`test_rmdir_on_regfile`

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        o = self.run_cephfs_shell_cmd("rmdir dumpfile")
+        print_output = o.stderr.getvalue()
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('Not a directory') != -1
```
same.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('Directory not empty') != -1
+
+    def test_rmdir_invalid_dir(self):
```
`test_rmdir_on_uncreated_dir`?

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        except CommandFailedError:
+            pass
+
+    def test_rmdir_p_invalid_dir(self):
```
same - `test_rmdir_p_uncreated_dir`

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        o = self.run_cephfs_shell_cmd("rmdir -p test_dir")
+        print_output = o.stderr.getvalue()
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('No such file or directory') != -1
```
same -- look for return value and actual absence/presence of files.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -162,6 +162,124 @@ def test_mkdir_path_with_path_option(self):
+        o = self.run_cephfs_shell_cmd("rmdir -p test_dir")
+        print_output = o.stderr.getvalue()
+        log.info("cephfs-shell output:\n{}".format(print_output))
+        assert print_output.find('Not a directory') != -1
```
same -- look for return value and actual absence/presence of files.

varshar16 commented:
@rishabh-d-dave I have removed all `CommandFailError` instances. Instead of looking for particular error message, I have changed it to just check if error is produced when non existing files or directories are removed.

batrick commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -175,6 +175,111 @@ def test_mkdir_path_with_path_option(self):
+            pass
+
+        self.assertTrue(path.exists(path.join(self.mount_a.mountpoint, "dumpfile")),
+                        "Something went wrong!! rmdir deleted a file")
```
The reason this works with `vstart_runner.py` and not teuthology is that `mount_a` mount is on the same machine as your the process executing these tests. When you run on teuthology, this python code is run separately from the ceph cluster. You need to execute a `stat` shell command or simliar on one of hte client mounts to check the existence of the file. See `stat` in mount.py.

batrick replied:
````
2020-01-13T09:56:30.261 INFO:tasks.cephfs_test_runner:======================================================================
2020-01-13T09:56:30.261 INFO:tasks.cephfs_test_runner:FAIL: test_rmdir_existing_file (tasks.cephfs.test_cephfs_shell.TestRMdir)
2020-01-13T09:56:30.261 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-01-13T09:56:30.261 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-01-13T09:56:30.261 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-cephfs-shell-fix-rmdir-p-option/qa/tasks/cephfs/test_cephfs_shell.py", line 238, in test_rmdir_existing_file
2020-01-13T09:56:30.262 INFO:tasks.cephfs_test_runner:    "Something went wrong!! rmdir deleted a file")
2020-01-13T09:56:30.262 INFO:tasks.cephfs_test_runner:AssertionError: Something went wrong!! rmdir deleted a file
````

varshar16 replied:
Thanks for explaining. I have updated the tests. Please review them again.

varshar16 commented:
Passed test results: http://pulpito.ceph.com/varsha-2020-01-20_13:17:03-fs-wip-cephfs-shell-fix-rmdir-p-option-distro-basic-smithi/4688138/

rishabh-d-dave commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -876,16 +876,21 @@ class CephFSShell(Cmd):
             if args.parent:
-                files = reversed(sorted(set(dirwalk(path))))
+                path = os.path.join(cephfs.getcwd(), path.rsplit(b'/')[0])
+                files = list(reversed(sorted(set(dirwalk(path)))))
```
calling`reversed` and `sorted` isn't much useful, AFAIS.

varshar16 replied:
True, I will update it.

rishabh-d-dave commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -876,16 +876,21 @@ class CephFSShell(Cmd):
+                    except libcephfs.Error as e:
+                        perror(e)
+                        path = b'.'
+                        break
```
Wouldn't the right behaviour be too continue the loop instead of break out of it? Breaking out of it would leave rest of the files undeleted.

And if one of the files couldn't be, ideally, the command should attempt to quit right away. Although, this does happen, but it would be better if code does that clearly. So setting a flag in the `execpt` block and then checking the flag and quitting (depending on flag) would be nicer and cleaner, IMO.

varshar16 replied:
In unix rmdir this does not happen, it throws error for the invalid file but deletes the rest.
```
[varsha@localhost testdir]$ mkdir -p t1/t2
[varsha@localhost testdir]$ ls
t1
[varsha@localhost testdir]$ rmdir -p inval_dir t1/t2
rmdir: failed to remove 'inval_dir': No such file or directory
[varsha@localhost testdir]$ ls
[varsha@localhost testdir]$ 
```

rishabh-d-dave replied:
> ```[varsha@localhost testdir]$ rmdir -p inval_dir t1/t2
rmdir: failed to remove 'inval_dir': No such file or directory
[varsha@localhost testdir]$ ls
[varsha@localhost testdir]$ 
```

Umm, I wasn't addressing this case. Anyways, I feel the point I was trying to make falls outside the scope of this PR. I'll go ahead approve. :)

varshar16 replied:
Sorry, I did not get your point. Can you please elaborate it again?

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -175,6 +175,86 @@ def test_mkdir_path_with_path_option(self):
         o = self.mount_a.stat('d5/d6/d7')
         log.info("mount_a output:\n{}".format(o))
 
+class TestRMdir(TestCephFSShell):
```
`Rmdir` would be preferable since we have a class named `Mkdir`.

varshar16 replied:
Done.

varshar16 commented:
Passed test results: http://pulpito.ceph.com/varsha-2020-01-22_16:39:45-fs-wip-cephfs-shell-fix-rmdir-p-option-distro-basic-smithi/
